### PR TITLE
test/mpi: Add support for GPU testing

### PR DESCRIPTION
**Pull request by raffenet**
## Pull Request Description

Integrate support for GPU (CUDA) testing. Memory allocation utilities are added. Proof of concept integration with pt2pt dtpools tests in `test/mpi/pt2pt`. Only send buffers are modified so far.

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by raffenet** [comment]
Rebased this on latest master. This version includes recv buffer modifications. pt2pt dtpools tests now exercise all the following send/recv pairs.
```
host->host (same as before)
host->registered
host->device
registered->host
registered->registered
registered->device
device->host
device->registered
device->device
```

**Comment by raffenet** [comment]
FYI, I am planning to modify the RMA tests next.

**Comment by pavanbalaji** [comment]
@raffenet  Some of the tests are timing out because of the extra large CPU-to-GPU copy code.  Should we reduce the number of iterations (testsize) for the large message tests?

**Comment by raffenet** [comment]
> @raffenet Some of the tests are timing out because of the extra large CPU-to-GPU copy code. Should we reduce the number of iterations (testsize) for the large message tests?

Sure, I will update the GPU tests for fewer iterations. I just added some RMA tests in a new commit. put, get and accumulate are covered so far.

**Comment by raffenet** [comment]
Updated more RMA tests and fixed some bugs. Origin and target buffers are modified for all tests. Result buffers not yet done.

**Comment by raffenet** [comment]
Added a commit modifying `get_accumulate.c` to put result buffers in device memory, but origin and target remain in host memory for now. I need to hack into the dtp test generation script a little bit more to add device result buffer testing in lock_x_dt.c, but not pollute the other RMA tests with all 27 origin,target,result buffer combinations.

**Comment by pavanbalaji** [comment]
@raffenet Can you increase the seed for each test, rather than keeping the seed the same for the different variants of the device-host combinations?  That'll test more datatype combinations.

File: test/mpi/maint/gentests_dtp.sh [hunk 1]
```diff
@@ -143,7 +144,15 @@ while read -r line ; do
...
+                    echo "${testname} $procs arg=-type=${type} arg=-sendcnt=${sendcount} arg=-recvcnt=${recvcount} arg=-seed=$seed arg=-testsize=${testsize} ${other_args} arg=-sendmem=reg_host arg=-recvmem=device $timelimit" >> ${builddir}/${testdir}/testlist.gpu
+                    echo "${testname} $procs arg=-type=${type} arg=-sendcnt=${sendcount} arg=-recvcnt=${recvcount} arg=-seed=$seed arg=-testsize=${testsize} ${other_args} arg=-sendmem=device arg=-recvmem=host $timelimit" >> ${builddir}/${testdir}/testlist.gpu
+                    echo "${testname} $procs arg=-type=${type} arg=-sendcnt=${sendcount} arg=-recvcnt=${recvcount} arg=-seed=$seed arg=-testsize=${testsize} ${other_args} arg=-sendmem=device arg=-recvmem=reg_host $timelimit" >> ${builddir}/${testdir}/testlist.gpu
+                    echo "${testname} $procs arg=-type=${type} arg=-sendcnt=${sendcount} arg=-recvcnt=${recvcount} arg=-seed=$seed arg=-testsize=${testsize} ${other_args} arg=-sendmem=device arg=-recvmem=device $timelimit" >> ${builddir}/${testdir}/testlist.gpu
```
Comment (pavanbalaji) [inline test/mpi/maint/gentests_dtp.sh]: I don't think we need so many combinations.  host-to-reghost and reghost-to-host are the same as host-to-host in mpich.  So the tests are basically equivalent.  The following combinations might be sufficient:

host-to-device
reghost-to-device
device-to-host
device-to-reghost
device-to-device

Even with those, mpich itself doesn't distinguish host vs. reghost.  That distinction only happens in yaksa, where it is already being tested.  So, perhaps, we could reduce the number of reghost tests?  Fewer iterations, perhaps?

**Comment by raffenet** [comment]
I'm still hacking on the test generation script to clean up the addition of the new tests. I've addressed the latest round of comments, except reducing iterations for the registered hosts tests. Will work on that in the morning.

**Comment by pavanbalaji** [comment]
Can host-to-host be removed from the GPU testlist?

**Comment by raffenet** [comment]
host-to-host should only be in testlist.dtp in this version.

**Comment by raffenet** [comment]
I removed the get_accumulate.c test modification commit, since there was no way to disable it when GPU support was not present. This can probably be reviewed and merged, and cleanups and improvements will come in follow up PRs.

**Comment by raffenet** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by pavanbalaji** [comment]
> I removed the get_accumulate.c test modification commit, since there was no way to disable it when GPU support was not present. This can probably be reviewed and merged, and cleanups and improvements will come in follow up PRs.

@raffenet Will you fix the CUDA configure integration first?

**Comment by pavanbalaji** [comment]
Some of the fixup commits can perhaps be squashed too before this is ready for review?

**Comment by raffenet** [comment]
> > I removed the get_accumulate.c test modification commit, since there was no way to disable it when GPU support was not present. This can probably be reviewed and merged, and cleanups and improvements will come in follow up PRs.
> 
> @raffenet Will you fix the CUDA configure integration first?

Can you be more specific?

**Comment by pavanbalaji** [comment]
> > > I removed the get_accumulate.c test modification commit, since there was no way to disable it when GPU support was not present. This can probably be reviewed and merged, and cleanups and improvements will come in follow up PRs.
> > 
> > 
> > @raffenet Will you fix the CUDA configure integration first?
> 
> Can you be more specific?

I'm referring to the below commit.  Will you fix it up to remove the hack?

```
test/mpi: Add configure check for CUDA support

INCOMPLETE: This is hacky in the way it adds CUDA dependencies to the
testsuite, but it will work for --with-cuda=<foo> on our Jenkins
nodes.
```

**Comment by raffenet** [comment]
> > > > I removed the get_accumulate.c test modification commit, since there was no way to disable it when GPU support was not present. This can probably be reviewed and merged, and cleanups and improvements will come in follow up PRs.
> > > 
> > > 
> > > @raffenet Will you fix the CUDA configure integration first?
> > 
> > 
> > Can you be more specific?
> 
> I'm referring to the below commit. Will you fix it up to remove the hack?
> 
> ```
> test/mpi: Add configure check for CUDA support
> 
> INCOMPLETE: This is hacky in the way it adds CUDA dependencies to the
> testsuite, but it will work for --with-cuda=<foo> on our Jenkins
> nodes.
> ```

Yes, I will work on that and squashing other fixups.

**Comment by pavanbalaji** [comment]
Is putting the result buffer on the GPU tricky in some way?  Can't that be handled in this PR too, so we can declare RMA tests as complete as well?

**Comment by raffenet** [comment]
It depends on which combinations we want to test. If the result buffer is the same memory type as the origin buffer, then it is trivial. If we want to test more combinations, then it will just take a little more time to get the test generation script right.

**Comment by raffenet** [comment]
Added a commit that includes support for result buffers and places them in the same memory type as origin buffers. The argument is just ignored by RMA tests that do not include `MPI_Get_accumulate`.

**Comment by raffenet** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by pavanbalaji** [comment]
> It depends on which combinations we want to test. If the result buffer is the same memory type as the origin buffer, then it is trivial. If we want to test more combinations, then it will just take a little more time to get the test generation script right.

We do want to test the combination that result buf is on a different type of memory than the origin.  What's the complexity?  Just the identification of which tests have Get_accumulate and which do not?  Would it be easier to generate all combinations, but have tests that do not have Get_accumulate simply return success for cases where the result buffer is anything but "host"?

**Comment by raffenet** [comment]
> > It depends on which combinations we want to test. If the result buffer is the same memory type as the origin buffer, then it is trivial. If we want to test more combinations, then it will just take a little more time to get the test generation script right.
> 
> We do want to test the combination that result buf is on a different type of memory than the origin. What's the complexity? Just the identification of which tests have Get_accumulate and which do not? Would it be easier to generate all combinations, but have tests that do not have Get_accumulate simply return success for cases where the result buffer is anything but "host"?

There's only one test source that has MPI_Get_accumulate within the DTPools tests (lock_x_dt.c). I was prototyping identifying MPI_Get_accumulate tests in `dtp-test-config.txt` and using that information to generate the combinations in `gentests_dtp.sh`. That was my idea, at least. I can include it in this PR once I finish it.

**Comment by raffenet** [comment]
> > > It depends on which combinations we want to test. If the result buffer is the same memory type as the origin buffer, then it is trivial. If we want to test more combinations, then it will just take a little more time to get the test generation script right.
> > 
> > 
> > We do want to test the combination that result buf is on a different type of memory than the origin. What's the complexity? Just the identification of which tests have Get_accumulate and which do not? Would it be easier to generate all combinations, but have tests that do not have Get_accumulate simply return success for cases where the result buffer is anything but "host"?
> 
> There's only one test source that has MPI_Get_accumulate within the DTPools tests (lock_x_dt.c). I was prototyping identifying MPI_Get_accumulate tests in `dtp-test-config.txt` and using that information to generate the combinations in `gentests_dtp.sh`. That was my idea, at least. I can include it in this PR once I finish it.

Which combinations should I include? For origin and target we have:
```
host-to-device
reghost-to-device
device-to-host
device-to-reghost
device-to-device
```
Maybe something like:
```
host-to-device-to-host
host-to-device-to-device
reg_host-to-device-reg_host
reg_host-to-device-device
device-to-host-to-host
device-to-host-to-device
device-to-reg_host-to-reg_host
device-to-reg_host-to-device
device-to-device-to-host
device-to-device-to-reg_host
device-to-device-to-device
```

**Comment by pavanbalaji** [comment]
> > > > It depends on which combinations we want to test. If the result buffer is the same memory type as the origin buffer, then it is trivial. If we want to test more combinations, then it will just take a little more time to get the test generation script right.
> > > 
> > > 
> > > We do want to test the combination that result buf is on a different type of memory than the origin. What's the complexity? Just the identification of which tests have Get_accumulate and which do not? Would it be easier to generate all combinations, but have tests that do not have Get_accumulate simply return success for cases where the result buffer is anything but "host"?
> > 
> > 
> > There's only one test source that has MPI_Get_accumulate within the DTPools tests (lock_x_dt.c). I was prototyping identifying MPI_Get_accumulate tests in `dtp-test-config.txt` and using that information to generate the combinations in `gentests_dtp.sh`. That was my idea, at least. I can include it in this PR once I finish it.
> 
> Which combinations should I include? For origin and target we have:
> 
> ```
> host-to-device
> reghost-to-device
> device-to-host
> device-to-reghost
> device-to-device
> ```
> 
> Maybe something like:
> 
> ```
> host-to-device-to-host
> host-to-device-to-device
> reg_host-to-device-reg_host
> reg_host-to-device-device
> device-to-host-to-host
> device-to-host-to-device
> device-to-reg_host-to-reg_host
> device-to-reg_host-to-device
> device-to-device-to-host
> device-to-device-to-reg_host
> device-to-device-to-device
> ```

I don't think we need that many combinations.  Perhaps the following are enough:

```
host-to-device-device
reghost-to-device-device
device-to-host-device
device-to-reghost-device
device-to-device-device
device-to-host-device
device-to-host-reghost
```

In any case, it doesn't really matter so much for this PR.  When we get more experience with the testing and how long it's taking, we can adjust it later.

**Comment by pavanbalaji** [comment]
@raffenet are any of these tests testing multiple devices?  Or are all of them on device 0?

**Comment by raffenet** [comment]
> @raffenet are any of these tests testing multiple devices? Or are all of them on device 0?

The device allocator uses a device id that is incremented after each allocation, so it uses multiple devices. That scheme was borrowed from yaksa.

**Comment by pavanbalaji** [comment]
> > @raffenet are any of these tests testing multiple devices? Or are all of them on device 0?
> 
> The device allocator uses a device id that is incremented after each allocation, so it uses multiple devices. That scheme was borrowed from yaksa.

May I suggest that the start device ID be set to `1` instead of `0`?  That would also test the case where the user is trying to move data from a device that is not its "current device" (in CUDA terminology).

**Comment by raffenet** [comment]
Made changes to the allocator and implemented result buffer testing in this version. I still need to test this JLSE before this is ready for another review.

**Comment by raffenet** [comment]
Need to fix the failures from the last run, too.

**Comment by raffenet** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by raffenet** [comment]
test:mpich/ch3/tcp

**Comment by raffenet** [comment]
test:mpich/ch3/tcp

**Comment by raffenet** [comment]
test:mpich/ch3/tcp

**Comment by raffenet** [comment]
test:mpich/ch3/tcp

**Comment by raffenet** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by pavanbalaji** [comment]
@raffenet  I think memory allocation failures should be fatal errors, and not just an increment of `err`.  We had to debug some nasty failures on solaris because the memory allocation had failed but the program still moved forward and segfaulted in an unrelated location.

**Comment by raffenet** [comment]
OK, I will make that change in the tests.

**Comment by raffenet** [comment]
FWIW, the centos32 failures in the latest runs do not occur if I change the seed value. I tested with the original seed, and also decrementing and incrementing the value that causes the failure. All pass. It's happening during DTPools buffer initialization.

**Comment by pavanbalaji** [comment]
> FWIW, the centos32 failures in the latest runs do not occur if I change the seed value. I tested with the original seed, and also decrementing and incrementing the value that causes the failure. All pass. It's happening during DTPools buffer initialization.

Changing the seed changes the datatypes that get tested.  It'll be good to retain the test/seed that's failing, at least as a separate test for debugging purposes.

**Comment by raffenet** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by raffenet** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by raffenet** [comment]
Recording failures from previous review run here for now. ch3/tcp configuration on 32-bit CentOS.
```
./pt2pt/sendrecv1 2 -type=MPI_INT:4+MPI_DOUBLE:8 -sendcnt=65530 -recvcnt=65530 -seed=301 -testsize=32 -sendmem=host -recvmem=host
 ./rma/lockall_dt_flushlocal 4 -type=MPI_INT -count=65530 -seed=1493 -testsize=16 -origmem=host -targetmem=host -resultmem=host
```

**Comment by raffenet** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by raffenet** [comment]
I don't think the ch3/tcp failures will be present in this version due to seed changes in the test generation script. I have made #4639 to investigate them.

**Comment by raffenet** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by raffenet** [comment]
Tests look like they will be clean this run.

File: test/mpi/pt2pt/pingping.c [hunk 2]
```diff
@@ -96,17 +101,18 @@ int main(int argc, char *argv[])
                 maxmsg = MAXMSG;
 
             if (rank == source) {
-                sendbuf = malloc(send_obj.DTP_bufsize);
-                if (sendbuf == NULL) {
+                MTestAlloc(send_obj.DTP_bufsize, sendmem, &sendbuf_h, &sendbuf);
+                if (sendbuf == NULL || sendbuf_h == NULL) {
```
Comment (pavanbalaji) [inline test/mpi/pt2pt/pingping.c]: Should be an assert

File: test/mpi/pt2pt/pingping.c [hunk 3]
```diff
@@ -129,9 +135,9 @@ int main(int argc, char *argv[])
                     }
                 }
 
-                free(sendbuf);
+                MTestFree(sendmem, sendbuf_h, sendbuf);
             } else if (rank == dest) {
-                recvbuf = malloc(recv_obj.DTP_bufsize);
+                MTestAlloc(recv_obj.DTP_bufsize, recvmem, &recvbuf_h, &recvbuf);
```
Comment (pavanbalaji) [inline test/mpi/pt2pt/pingping.c:138]: Should be an assert

File: test/mpi/pt2pt/sendrecv1.c [hunk 4]
```diff
@@ -86,17 +91,18 @@ int main(int argc, char *argv[])
             }
 
             if (rank == source) {
-                sendbuf = malloc(send_obj.DTP_bufsize);
-                if (sendbuf == NULL) {
+                MTestAlloc(send_obj.DTP_bufsize, sendmem, &sendbuf_h, &sendbuf);
+                if (sendbuf == NULL || sendbuf_h == NULL) {
```
Comment (pavanbalaji) [inline test/mpi/pt2pt/sendrecv1.c]: Should be an assert

File: test/mpi/pt2pt/sendrecv1.c [hunk 5]
```diff
@@ -110,19 +116,20 @@ int main(int argc, char *argv[])
                     }
                 }
 
-                free(sendbuf);
+                MTestFree(sendmem, sendbuf_h, sendbuf);
             } else if (rank == dest) {
-                recvbuf = malloc(recv_obj.DTP_bufsize);
-                if (recvbuf == NULL) {
+                MTestAlloc(recv_obj.DTP_bufsize, recvmem, &recvbuf_h, &recvbuf);
+                if (recvbuf == NULL || recvbuf_h == NULL) {
```
Comment (pavanbalaji) [inline test/mpi/pt2pt/sendrecv1.c]: Should be an assert

File: test/mpi/maint/gentests_dtp.sh [hunk 6]
```diff
@@ -158,6 +158,38 @@ while read -r line ; do
                     echo "${testname} $procs arg=-type=${type} arg=-sendcnt=${sendcount} arg=-recvcnt=${recvcount} arg=-seed=$seed arg=-testsize=${gputestsize} ${other_args} arg=-sendmem=device arg=-recvmem=device $timelimit" >> ${builddir}/${testdir}/testlist.gpu
                     seed=$((seed + 1))
                 done
+            elif [ $testdir = "rma" ] ; then
+                if grep -q ^lock <<<"$testname" ; then # lock_x_dt.c uses MPI_Get_accumulate, so we must specify resultmem
```
Comment (pavanbalaji) [inline test/mpi/maint/gentests_dtp.sh:162]: This code logic seems very fragile.  Can you at least check for the full file name?  Or better yet, create separate lists for tests that use Get_accumulate and tests that don't?

File: test/mpi/maint/gentests_dtp.sh [hunk 7]
```diff
@@ -190,6 +190,13 @@ while read -r line ; do
                     echo "${testname} $procs arg=-type=${type} arg=-count=${sendcount} arg=-seed=$seed arg=-testsize=${gputestsize} ${other_args} arg=-origmem=device arg=-targetmem=device $timelimit" >> ${builddir}/${testdir}/testlist.gpu
                     seed=$((seed + 1))
                     fi
+            elif [ $testdir = "coll" ] ; then
+                echo "${testname} $procs arg=-type=${type} arg=-count=${sendcount} arg=-seed=$seed arg=-testsize=${testsize} ${other_args} arg=-memtype=host $timelimit" >> ${builddir}/${testdir}/testlist.dtp
+                seed=$((seed + 1))
+                echo "${testname} $procs arg=-type=${type} arg=-count=${sendcount} arg=-seed=$seed arg=-testsize=${gputestsize} ${other_args} arg=-memtype=reg_host $timelimit" >> ${builddir}/${testdir}/testlist.gpu
+                seed=$((seed + 1))
+                echo "${testname} $procs arg=-type=${type} arg=-count=${sendcount} arg=-seed=$seed arg=-testsize=${gputestsize} ${xother_args} arg=-memtype=device $timelimit" >> ${builddir}/${testdir}/testlist.gpu
+                seed=$((seed + 1))
```
Comment (pavanbalaji) [inline test/mpi/maint/gentests_dtp.sh:199]: For the collectives, do we not want to generate cases where some processes use one type of memory and some processes don't?  How about modifying the tests so the memtype argument is used as "memtype + rank" by each rank?

File: test/mpi/pt2pt/pingping.c (hunk 2 above)
Comment (pavanbalaji) [reply test/mpi/pt2pt/pingping.c]: Oh, this is removed in a later commit.  Nevermind.

File: test/mpi/maint/gentests_dtp.sh (hunk 7 above)
Comment (raffenet) [reply test/mpi/maint/gentests_dtp.sh:199]: I will add this and re-submit.

File: test/mpi/maint/gentests_dtp.sh (hunk 6 above)
Comment (raffenet) [reply test/mpi/maint/gentests_dtp.sh:162]: Will create a list for get_accumulate tests.

**Comment by raffenet** [comment]
This version adds an explicit parameter to `dtp-test-config.txt` to identify `MPI_GET_ACCUMULATE` tests. Since we are already reading from this config file, it seemed more appropriate than adding lists to the test generation script. For the collectives tests, I made buffer type dependent on odd or even rank. The following combinations are included in this version, but can be easily updated:
```
even: host, odd: device
even: device, odd: host
even: device, odd: device
```
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by pavanbalaji** [comment]
> ```
> even: host, odd: device
> even: device, odd: host
> even: device, odd: device
> ```

Can the second one be made device to registered host, to cover a few more cases?

**Comment by raffenet** [comment]
> > ```
> > even: host, odd: device
> > even: device, odd: host
> > even: device, odd: device
> > ```
> 
> Can the second one be made device to registered host, to cover a few more cases?

Done.

**Pull request merged**